### PR TITLE
SubteamCreateEngine

### PR DESCRIPTION
**Pull request by oconnor663**
r? @maxtaco @mlsteele 

As with `TeamCreateEngine`, right now I've written it to just make the creator into the sole admin of the group. We can plumb through more control over that as we build UI for it?

### CODE REVIEW

File: go/teams/sig.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	return ret, nil
+}
+
+// 15 random bytes, followed by the byte 0x25, encoded as hex
+func NewSubteamID() keybase1.TeamID {
+	randBytes, err := libkb.RandBytes(15)
+	if err != nil {
+		panic("RandBytes failed: " + err.Error())
+	}
+	return keybase1.TeamID(hex.EncodeToString(randBytes) + "25")
```
Comment (maxtaco) [inline go/teams/sig.go]: make a constants file constant for this.

File: go/teams/sig.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+		Reader []libkb.NameWithEldestSeqno `json:"reader"`
+	} `json:"members"`
+	PerTeamKey struct {
+		Generation    int          `json:"generation"`
+		EncryptionKID keybase1.KID `json:"encryption_kid"`
+		SigningKID    keybase1.KID `json:"signing_kid"`
+		// reverse_sig always gets set to null, and the caller has to overwrite it afterwards
+	} `json:"per_team_key"`
+	Parent *ParentSection `json:"parent,omitempty"`
+}
```
Comment (maxtaco) [inline go/teams/sig.go]: use existing structs

File: go/teams/sig.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	body := ret.AtKey("body")
+	body.SetKey("team", teamSectionJSON)
+
+	return ret, nil
+}
+
+// the first 15 bytes of the sha256 of the lowercase team name, followed by the byte 0x24, encoded as hex
+func RootTeamIDFromName(name string) keybase1.TeamID {
+	sum := sha256.Sum256([]byte(strings.ToLower(name)))
+	return keybase1.TeamID(hex.EncodeToString(sum[0:15]) + "24")
```
Comment (maxtaco) [inline go/teams/sig.go]: use constant

File: go/teams/sig.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	subteamSection := jsonw.NewDictionary()
+	subteamSection.SetKey("id", jsonw.NewString(string(subteamID)))
+	subteamSection.SetKey("name", jsonw.NewString(subteamFQName))
+	teamSection := jsonw.NewDictionary()
+	teamSection.SetKey("id", jsonw.NewString(string(parentTeam.ID)))
+	teamSection.SetKey("subteam", subteamSection)
+	body := ret.AtKey("body")
+	body.SetKey("team", teamSection)
```
Comment (maxtaco) [inline go/teams/sig.go]: use `wrapperFromObject` here?

File: go/engine/subteam_create.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+		Sig:        v2Sig,
+		SigningKID: signingKey.GetKID(),
+		Type:       string(libkb.LinkTypeNewSubteam),
+		SigInner:   string(newSubteamSigJSON),
+		TeamID:     parentTeam.GetID(),
+	}
+	return
+}
+
+func (e *SubteamCreateEngine) generateHeadSigForSubteamChain(ctx *Context, me *libkb.User, signingKey libkb.GenericKey, parentTeam *teams.TeamSigChainState) (item *libkb.SigMultiItem, boxes *PerTeamSharedSecretBoxes, err error) {
```
Comment (maxtaco) [inline go/engine/subteam_create.go]: how similar is this to what happens for a Root team and can we reuse and existing code?

File: go/engine/subteam_create.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package engine
```
Comment (maxtaco) [inline go/engine/subteam_create.go]: do we want a `engine_team` subfolder?

**Comment by oconnor663** [comment]
CI failure is an import cycle in the go/teams test build, will fix.

File: go/kbtest/kbtest.go [hunk 7]
```diff
@@ -70,3 +87,128 @@ func CreateAndSignupFakeUser(prefix string, g *libkb.GlobalContext) (*FakeUser,
...
+
+func (u *testProvisionUI) ProvisioneeSuccess(_ context.Context, _ keybase1.ProvisioneeSuccessArg) error {
+	u.printf("ProvisioneeSuccess")
+	return nil
+}
+
+func (u *testProvisionUI) ProvisionerSuccess(_ context.Context, _ keybase1.ProvisionerSuccessArg) error {
+	u.printf("ProvisionerSuccess")
+	return nil
+}
```
Comment (oconnor663) [inline go/kbtest/kbtest.go]: This file has gained a lot of duplicated code from `engine/common_test.go`, especially the `testProvisionUI` bit. @patrickxb, do you think it's worth lowering some of that down from engine to a sharable test module?

**Comment by oconnor663** [comment]
@maxtaco ready for another review.

File: go/kbtest/kbtest.go (hunk 7 above)
Comment (patrickxb) [inline go/kbtest/kbtest.go]: you could probably use a much smaller ProvisionUI here.  PromptNewDeviceName is probably the only thing that needs to actually return something useful.  There might be a smaller one somewhere else?

Comment (patrickxb) [inline go/kbtest/kbtest.go]: I hesitate to change anything with provisioning (and its tests).

If you'd like to reuse the provision ui in the engine test, you could just make the test there?

**Pull request merged**